votor/src/consensus_pool_service.rs

```diff
@@ -412,3 +412,269 @@ impl ConsensusPoolService {
+                assert_eq!(*slot, 2);
+                assert_eq!(*receivied_block_id, block_id);
+                assert!(*is_fast_finalized);
+                true
+            } else {
+                false
+            }
+        });
+        // Verify that we received a commitment update
+        wait_for_event(
+            &setup_result.commitment_receiver,
+            |AlpenglowCommitmentAggregationData {
+                 commitment_type,
+                 slot,
+                 ..
+             }| {
+                assert_eq!(*commitment_type, AlpenglowCommitmentType::Finalized);
+                assert_eq!(*slot, 2);
+                true
+            },
+        );
+
+        // Now send a Skip certificate on slot 3, should be forwarded immediately
+        let skip_certificate = CertificateMessage {
+            certificate: Certificate::Skip(3),
+            signature: BLSSignature::default(),
+            bitmap: vec![],
+        };
+        consensus_message_sender
+            .send(ConsensusMessage::Certificate(skip_certificate))
+            .unwrap();
+        wait_for_event(&setup_result.bls_receiver, |event| {
+            if let BLSOp::PushCertificate { certificate } = event {
+                if matches!(certificate.certificate, Certificate::Skip(3)) {
+                    return true;
+                }
+            }
+            false
+        });
+        setup_result.exit.store(true, Ordering::Relaxed);
+        consensus_pool_service.join().unwrap();
+    }
+
+    #[test]
+    fn test_produce_block_event() {
+        let setup_result = setup();
+        // Find when is the next leader slot for me (validator 0)
+        let my_pubkey = setup_result.validator_keypairs[0].node_keypair.pubkey();
+        let next_leader_slot = setup_result
+            .leader_schedule_cache
+            .next_leader_slot(&my_pubkey, 0, &setup_result.root_bank.load(), None, 1000000)
+            .expect("Should find a leader slot");
+        // Send skip certificates for all slots up to the next leader slot
+        for slot in 1..next_leader_slot.0 {
+            let skip_certificate = CertificateMessage {
+                certificate: Certificate::Skip(slot),
+                signature: BLSSignature::default(),
+                bitmap: vec![],
+            };
+            setup_result
+                .consensus_message_sender
+                .send(ConsensusMessage::Certificate(skip_certificate))
+                .unwrap();
+        }
+        // Verify that we received a produce block event
+        wait_for_event(&setup_result.event_receiver, |event| {
+            if let VotorEvent::ProduceWindow(LeaderWindowInfo { start_slot, .. }) = event {
+                assert_eq!(*start_slot, next_leader_slot.0);
+                true
+            } else {
+                false
+            }
+        });
+        setup_result.exit.store(true, Ordering::Relaxed);
```

Comment:
Could we make these send -> receive tests more generic by handing a function `test_sender_receiver()` a set of (sender, receiver, event) and then checking that the receiver gets the event?

We can send construct an array of correct/expected (sender, receiver, event) combinations and loop through it running that function as a sanity check